aea/aea_builder.py

```diff
@@ -0,0 +1,164 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2020 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This module contains utilities for building an AEA."""
+from typing import Dict, Any, Tuple, Set
+
+from aea.aea import AEA
+from aea.configurations.base import PublicId, PackageConfiguration, AgentConfig, ConfigurationType
+from aea.configurations.loader import ConfigLoader
+from aea.crypto.ledger_apis import LedgerApis
+from aea.crypto.wallet import Wallet
+from aea.identity.base import Identity
+from aea.mail.base import Address
+from aea.protocols.base import Protocol
+from aea.registries.base import Resources
+
+PackageId = Tuple[AgentComponentType, PublicId]
+
+
+class _Package:
+
+    def __init__(self, package_type, package_configuration):
+        """"""
+
+        self._package_type = package_type  # one of protocol, skill, connection etc.
+        self._package_config = package_configuration
+        self._modules = {}
+
+    @classmethod
+    def load(self, package_type) -> "_Package":
+        """Load package from dir"""
+        # read config file
+        # load modules
+
+
+class AEABuilder:
+    """This class helps to build an AEA."""
+
+    def __init__(self):
+        self._wallet = Wallet({})  # TODO make Wallet mutable (add/remove_private_key)
+        self._ledger_apis = LedgerApis({}, "")  # TODO make it mutable (add/remove api)
+
+        self._resources = Resources()
+        self._connections = []
+
+        # identity
+        self._name = None
+        self._addresses = {}  # type: Dict[str, Address]
+        self._default_key = None  # set by the user, or instantiate a default one.
+
+        # from package id to packages it depends on.
+        # we should detect loops at "add" time, by inspecting the component config file.
+        self._dependency_graph = {}  # type: Dict[PackageId, Set[_Package]]
+
+        # add default protocol
```

Comment:
I think the builder should be adding the default packages during initialisation

Reply:
OK, I missed this comment for some reason.
The current approach lets the user decide if the `AEABuilder` should be initialized with the default packages or not. Look at `__init__` method.